doc/pins/p4rt_app_read_cache_hld.md

```diff
@@ -0,0 +1,100 @@
+# P4RT Read Cache HLD
+
+
+## Revision
+
+Rev  | RevDate    | Author(s) | Change Description
+---- | ---------- | --------- | ------------------
+v0.1 | 02/06/2023 | Google    | Initial Version
+
+
+## Definitions/Abbreviations
+
+**P4**: Programming Protocol-independent Packet Processors (P4) is a domain-specific language for network devices, specifying how data plane devices (switches, NICs, routers, filters, etc.) process packets.
+
+**P4RT**: P4 Runtime (P4RT) is a control plane specification for controlling the data plane elements of a device defined or described by a P4 program.
+
+**PI**: Platform Independent representation of P4RT messages, which does not change if the P4 program changes.
+
+## Overview
+
+A P4 Runtime server provides multiple services for a controller to interact with the switch. However, during normal operation we expect most controllers will only rely on 3:
+
+*   Stream - to ensure the connection is healthy, and send/receive PacketIO requests.
+*   Write - to update forwarding state.
+*   Read - to verify forwarding state.
+
+Today the read performance can be significantly slower than the other operations, and because we only handle one outstanding request at a time this performance can affect other operations due to lock contention.
+
+
+## P4RT Cache Design
+
+The P4RuntimeImpl will maintain a table entry cache with values already stored in the PI format. This will eliminate any translations.
+
+*   absl::flat\_hash\_map\<TableEntryKey, p4::v1::TableEntry> table\_entry\_cache\_
+
+The cache will be updated for each P4RT App write request.
+
+![Read Cache Sequence Diagram](images/p4rt_app_read_cache_sequence_diagram.svg)
+
+**Write Operation** | **Status** | **Changes to the cache**
+------------------- | ---------- | ------------------------
+\<any\>             | FAIL       | Do nothing
+INSERT              | SUCCESS    | Create a new PI entry
+MODIFY              | SUCCESS    | Update an existing PI entry
+DELETE              | SUCCESS    | Remove an existing PI entry
+
+While warmboot is not currently supported in the P4RT App. When it is, we should be able to read any data back from Redis (i.e. same as we do today) and use those values to pre-populate the cache before warmboot completes.
+
+To ensure the contents of the cache are in sync with Redis we will include verification logic that reads entries from Redis, translates them to PI and compares them to the cache. Operators can choose how often to run these checks.
```

Comment:
I added some clarifications around what will and will not be checked.

The contract between P4RT App and the P4Orch is that updates only happen when the controller issues a request. This is needed since it's up to the controller to decide how long it waits between reads. With that in mind any change between the cache and AppDb is going to be a much bigger concern, and we don't want to mask it by purging entries or trying to get them in sync.

Practically speaking this verification logic will be most useful during development to ensure we don't break the contract. In deployments we really shouldn't see a difference so operators can run it every hour, every few hours, once or twice a day, etc.

Reply:
sure, and I see that you have updated design to report inconsistency (and not fix any inconsistency as a result of issued checker), which is fine since controllers/operators can choose to fix inconsistecies it via other means.